mm2src/lp_swap/recover_swap.rs

```diff
@@ -0,0 +1,342 @@
+        // Don't set the fee since the value is used when we calculate locked by other swaps amount only.
+        fee_to_send_taker_fee: None,
+        // Don't set the fee since the value is used when we calculate locked by other swaps amount only.
+        taker_payment_trade_fee: None,
+        // Don't set the fee since the value is used when we calculate locked by other swaps amount only.
+        maker_payment_spend_trade_fee: None,
+        maker_coin_swap_contract_address: negotiated_event.maker_coin_swap_contract_addr.clone(),
+        taker_coin_swap_contract_address: negotiated_event.taker_coin_swap_contract_addr.clone(),
+    });
+    taker_swap.events.push(TakerSavedEvent {
+        timestamp: started_event_timestamp,
+        event: taker_started_event,
+    });
+
+    let taker_negotiated_event = TakerSwapEvent::Negotiated(MakerNegotiationData {
+        maker_payment_locktime: started_event.maker_payment_lock,
+        maker_pubkey: started_event.my_persistent_pub,
+        secret_hash: started_event.secret_hash.or_mm_err(|| RecoverSwapError::NoSecretHash)?,
+        maker_coin_swap_contract_addr: negotiated_event.maker_coin_swap_contract_addr,
+        taker_coin_swap_contract_addr: negotiated_event.taker_coin_swap_contract_addr,
+    });
+    taker_swap.events.push(TakerSavedEvent {
+        timestamp: negotiated_timestamp,
+        event: taker_negotiated_event,
+    });
+
+    // Then we can continue to process success Maker events.
+    taker_swap.events.extend(convert_maker_to_taker_events(event_it));
+
+    Ok(taker_swap)
+}
+
+/// Converts `MakerSwapEvent` to `TakerSwapEvent`.
+/// Please note that this method ignores the following events:
+/// * [`MakerSwapEvent::Started`] and [`MakerSwapEvent::Negotiated`]
+///   since they are used outside of this function to generate `TakerSwap` and the initial [`TakerSwapEvent::Started`] and [`TakerSwapEvent::Negotiated`] events.
+/// * [`TakerSwapEvent::TakerPaymentReceived`], [`TakerSwapEvent::TakerPaymentSpent`], [`TakerSwapEvent::MakerPaymentRefunded`]
+///   because we can't guarantee that `MakerPayment` is valid.
+fn convert_maker_to_taker_events(event_it: impl Iterator<Item = MakerSavedEvent>) -> Vec<TakerSavedEvent> {
+    let mut events = Vec::new();
+    for MakerSavedEvent { event, timestamp } in event_it {
+        match event {
+            MakerSwapEvent::TakerFeeValidated(tx_ident) => events.push(TakerSavedEvent {
+                timestamp,
+                event: TakerSwapEvent::TakerFeeSent(tx_ident),
+            }),
+            // Please note we have not to push `MakerPaymentValidatedAndConfirmed` since we could actually decline it.
+            MakerSwapEvent::MakerPaymentSent(tx_ident) => events.push(TakerSavedEvent {
+                timestamp,
+                event: TakerSwapEvent::MakerPaymentReceived(tx_ident),
+            }),
+            // We should ignore the `TakerPaymentReceived`, `TakerPaymentSpent`, `MakerPaymentRefunded` events
```

Comment:
This statement doesn't look correct. Taker sends his payment only after validation and confirmation of the maker's. So if the maker has a `TakerPaymentReceived` event then the taker had `MakerPaymentValidatedAndConfirmed` for sure.

Reply:
I agree with you, but if only the input swap file is trusted.
But I can imagine the only one case when we shouldn't trust external swap data is when the swap is ongoing.
Maybe we should prohibit the import of such ongoing swaps?
Anyway, it seems to be an out of this topic.

Regarding `recover_funds_of_swap`, the recovered swap files allow recovering own/spend other payment.
Do you think we should handle these `TakerPaymentReceived`, `TakerPaymentSpent`, `MakerPaymentRefunded` events?

Reply:
> I agree with you, but if only the input swap file is trusted.

The current expected use case is to make it easier for the API and support team to recreate lost swaps data. It should be validated "manually" for now. We should also mention in the docs that this RPC must not be used with the input provided by an untrusted third party. The same also implies to `import_swaps`.

> Do you think we should handle these TakerPaymentReceived, TakerPaymentSpent, MakerPaymentRefunded events?

I think we should map them as follows:
`TakerPaymentReceived` => `TakerPaymentSent`
`TakerPaymentSpent` => `TakerPaymentSpent`
`MakerPaymentRefunded` => `MakerPaymentSpendFailed`

Reply:
Will do!

Reply:
Done